src/main/java/org/jvnet/hudson/tools/versionnumber/VersionNumberBuilder.java

```diff
@@ -39,25 +41,35 @@
  * <p>
  * This plugin keeps track of its version through a {@link VersionNumberAction} attached
  * to the project.  Each build that uses this plugin has its own VersionNumberAction,
- * and this contains the builds today/this month/ this year/ all time.  When incrementing
+ * and this contains the builds today / this month / this year / all time.  When incrementing
  * each of these values, unless they're overridden in the configuration the value from
  * the previous build will be used.
  * </p>
+ * <p>
+ * Such a value can be either overridden with a plain number or with an environment-variable.
+ * In the later case the value will be read from the environment-variable at build-time. If
+ * it cannot be parsed as an integer the value from the previous build will be incremented
+ * and used instead.
+ * </p>
  *
  * @author Carl Lischeske - NETFLIX
+ * @author Deniz Bahadir - BENOCS
  */
 public class VersionNumberBuilder extends BuildWrapper {
     
     private static final DateFormat defaultDateFormat = new SimpleDateFormat("yyyy-MM-dd");
+    // Pattern:   ${VAR_NAME} or $VAR_NAME
+    private static final String envVarPattern = new String("^(?:\\$\\{(\\w+)\\})|(?:\\$(\\w+))$");
+
     private final String versionNumberString;
     private final Date projectStartDate;
     private final String environmentVariableName;
     private final String environmentPrefixVariable;
     
-    private int oBuildsToday;
-    private int oBuildsThisMonth;
-    private int oBuildsThisYear;
-    private int oBuildsAllTime;
+    private String oBuildsToday;
+    private String oBuildsThisMonth;
+    private String oBuildsThisYear;
+    private String oBuildsAllTime;
```

Comment:
Actually, this should not cause a compatibility issue when a user upgrades (I took care of this) but only if he tries to downgrade again. (But I do not think this must be our concern.)

The reason that it does not cause a compatibility issue is the following:

The original (int-)version of the plugin used negative numbers to say "just increase this counter by 1". If the user wanted to reset such a counter for the next build he had to store a positive number (or zero) which were stored in `config.xml` until the next build. (Then they were applied and replaced by -1 in `config.xml`.)

When the plugin is updated to the version with my changes, the values in `config.xml` are not modified. However, when the next build is triggered and these values are read, negative values will be interpreted as _invalid_ and replaced by an _empty string_ (due to method `makeValid`) while positive numbers will be interpreted as positive numbers.
As a result, everything works as before.

And yes, of course, I tested this.

(BTW: This is the reason for this second pull-request which forbids negative numbers.)